joblib/_parallel_backends.py

```diff
@@ -0,0 +1,324 @@
+        return
+
+    @abc.abstractmethod
+    def batch_completed(self, batch_size, duration):
+        """ Callback indicate how long it took to run a batch """
+        return
+
+    @abc.abstractmethod
+    def terminate(self):
+        """ Shutdown the process or thread pool """
+        return
+
+    @abc.abstractmethod
+    def apply_async(self, func, callback=None):
+        """ Schedule a func to be run """
+        return
+
+    @abc.abstractmethod
+    def get_exceptions(self):
+        return
+
+
+###############################################################################
+class SequentialBackend(ParallelBackendBase):
+    """A ParallelBackend which will execute all batches sequentially.
+    Does not use/create any threading objects, and hence has virtually no
+    overhead. Used when n_jobs == 1."""
+
+    def effective_n_jobs(self, n_jobs):
+        """ Determine the number of jobs which are going to run in parallel """
+        if n_jobs == 0:
+            raise ValueError('n_jobs == 0 in Parallel has no meaning')
+        return 1
+
+    def initialize(self, n_jobs, poolargs):
+        """ Build a process or thread pool and return the number of workers """
+        return 1
+
+    def compute_batch_size(self):
+        """ Determine the optimal batch size """
+        return 1
+
+    def batch_completed(self, batch_size, duration):
+        """ Callback indicate how long it took to run a batch """
+        pass
+
+    def terminate(self):
+        """ Shutdown the process or thread pool """
+        pass
+
+    def apply_async(self, func, callback=None):
+        """ Schedule a func to be run """
+        result = ImmediateResult(func)
+
+        if callback:
+            callback(result)
+
+        return result
+
+    def get_exceptions(self):
+        return []
+
+
+###############################################################################
+class PoolBackend(object):
+    """A helper class which defines some common methods for
+    Pool-like Parallel backends"""
+    __metaclass__ = abc.ABCMeta
+
+    def effective_n_jobs(self, n_jobs):
+        """ Determine the number of jobs which are going to run in parallel """
+        if n_jobs == 0:
+            raise ValueError('n_jobs == 0 in Parallel has no meaning')
+        elif mp is None or n_jobs is None:
+            # multiprocessing is not available or disabled, fallback
+            # to sequential mode
+            return 1
+        elif n_jobs < 0:
+            n_jobs = max(mp.cpu_count() + 1 + n_jobs, 1)
+        return n_jobs
+
+    def terminate(self):
+        """ Shutdown the process or thread pool """
+        if self._pool is not None:
+            self._pool.close()
+            self._pool.terminate()  # terminate does a join()
+            self._pool = None
+
+    def apply_async(self, func, callback=None):
+        """ Schedule a func to be run """
+        return self._pool.apply_async(SafeFunction(func), callback=callback)
```

Comment:
To be clear, Joblib is going to call this function not with the user supplied function, but with a function that maps the user-supplied function, correct?